elixir-sdk/native/sdk_core/src/conversion.rs

```diff
@@ -0,0 +1,56 @@
+use rustler::{Encoder, Env, NifResult, Term};
+use std::collections::HashMap;
+use std::sync::Arc;
+use eppo_core::{AttributeValue, Str, events::AssignmentEvent, ufc::AssignmentValue};
+use rustler::types::atom;
+
+
+pub fn convert_attributes(subject_attributes: Term) -> NifResult<Arc<HashMap<Str, AttributeValue>>> {
+    // Obtain an iterator over the map's key-value pairs.
+    let map: HashMap<String, Term> = subject_attributes.decode()?;
+    let mut attributes = HashMap::with_capacity(map.len());
+
+    for (key, value_term) in map {
+        // Try to decode the value as one of the supported types.
+        let attr_value = if let Ok(b) = value_term.decode::<bool>() {
+            // Booleans are stored as categorical attributes.
+            AttributeValue::categorical(b)
+        } else if let Ok(i) = value_term.decode::<i64>() {
+            // Integers are converted to f64 and stored as numeric.
+            AttributeValue::numeric(i as f64)
+        } else if let Ok(f) = value_term.decode::<f64>() {
+            AttributeValue::numeric(f)
+        } else if let Ok(s) = value_term.decode::<String>() {
+            // Strings are stored as categorical attributes.
+            AttributeValue::categorical(s)
```

Comment:
minor: it looks like `&str` implements `Decode`, and its very likely to be more efficient than allocating a new `String`. `Str` (and thus `AttributeValue`) also has a couple of optimizations for small strings to avoid copying, so `&X -> &str -> Str` is more preferable than `&X -> String -> Str`

```suggestion
        } else if let Ok(s) = value_term.decode::<&str>() {
            // Strings are stored as categorical attributes.
            AttributeValue::categorical(s)
```

Same also goes for map keys

Reply:
Sweet!